drivers: Add retained memory driver API

nordicjm opened the pull request:
Adds a retained memory driver interface with 2 backends (Nordic GPREGRET and RAM) which allows for storing data persistently across a reboot

- [x] Rename to retained-mem
- [x] Drop is valid function
- [x] Drop mcp SRAM (optionally submit as bbram driver)

bjarki-trackunit commented on include/zephyr/drivers/retention.h:
```diff
@@ -0,0 +1,195 @@
+ * @brief	Callback API to read retained data.
+ * See auxdisplay_custom_command() for argument description
+ */
+typedef int32_t (*retention_read_api)(const struct device *dev, size_t offset, uint8_t *buffer,
```
Using POSIX read() as a reference, reading data to a buffer takes this form
```
ssize_t read(int fd, void *buf, size_t count);
```
returning the number of bytes written to the buffer (or < 0 if error), instead of reusing the size argument. Like this:
```
typedef int32_t (*retention_read_api)(const struct device *dev, size_t offset, uint8_t *buffer,
                                      const size_t size);
```
It is a bit safer and easier to read since the size argument is const

nordicjm replied:
Changed to only allow reading and writing to valid locations without returning bytes read/written

bjarki-trackunit commented on include/zephyr/drivers/retention.h:
```diff
@@ -0,0 +1,195 @@
+ * See auxdisplay_custom_command() for argument description
+ */
+typedef int32_t (*retention_clear_api)(const struct device *dev);
+
```
It seems the implementation of this API is often just memset() to a known value. Unless there is hardware which has some bit or built in operation to set all memory to some preset value, or otherwise mark data as cleared, I would instead replace this API with a memset wrapper instead:
```
typedef int32_t (*retention_memset_api)(const struct device *dev, size_t offset, uint8_t value, size_t size);
```

nordicjm replied:
Added clear() function but unimplemented in current drivers as hardware does not support it

bjarki-trackunit commented on include/zephyr/drivers/retention.h:
```diff
@@ -0,0 +1,195 @@
+ *		if enabled).
+ * See auxdisplay_custom_command() for argument description
+ */
+typedef int32_t (*retention_is_valid_api)(const struct device *dev, bool *valid);
```
This function could potentially return 1 of data is valid, 0 if it is not, and < 0 if error.

nordicjm replied:
Updated

bjarki-trackunit commented on drivers/retention/retention_mcp.c:
```diff
@@ -0,0 +1,164 @@
+static int32_t microchip_mcp7940n_retention_set(const struct device *dev, size_t offset,
+						uint8_t *buffer, size_t size)
+{
+	const struct microchip_mcp7940n_retention_config *config = dev->config;
```
Note which affects all the drivers: modifying the memory should be protected from concurrent access from multiple threads

Laczen commented:
@nordicjm, thanks for this proposal. I am not quite sure why the prefix and the verification is added. The verification could be left to the user by using there prefered verification method (a crc, a magic, ...). The use of the prefix is not clear.

At the moment I'm doing something similar (for sram) by adding the following to the dts:
```
/ {
	retention {
		#address-cells = < 0x1 >;
		#size-cells = < 0x1 >;

		retention: sram@2001FFE0 {
			compatible = "zephyr,memory-region", "mmio-sram";
			reg = < 0x2001FFE0 0x0020 >;
			zephyr,memory-region = "RETENTION";
		};
	};
}; 
```

It is good to provide a special api to write/read this area. I would prefer an interface similar to flash. I wouldn't go for the posix standards that returns the written/read size.

nordicjm commented:
Prefix is for data to use at the beginning of the area to know that it is valid, likewise verification is a checksum that can be added to the end - both can be used, either can be used or neither can be used. The use cases:

- Prefix: Allows knowing if valid data has been set, if this is not set, it can be ignored, or it can be used by multiple different systems/images with different prefixes for different reasons
- Verification (checksum): Allows knowing if the data is valid and was written completely before the device was rebooted or power cycled, if there is a checksum mismatch, it can be ignored

I will move both of these (and the verification) to the common code instead of implementation specific code.

That area is essentially the same as how I've implemented it, I would prefer a way of it reserving the memory itself without the user having to do it (i.e. decrementing amount of RAM that application can use) but for now it'll just have to work like that.

I prefer the name clear rather than memset. I will need to revisit the read because I will probably remove returning the read byte size, if someone tries to read too much, it can just return an error, and the currently missing function (get details) will return details including the size (available to the user i.e. without prefix or verification bytes) that the end application can use.

Note that this will be brought up in the architecture working group meeting today

Laczen commented:
@nordicjm, so the prefix would be some kind of "magic" that can be used to determine the usage of the area and the crc to make sure it is valid. While I understand and support this as a good way of using the retention area it might be better to leave this up to the user. It is very easy for the user to define the structure of the area to suit his/her needs and to include a prefix/crc as needed without fixing it.

It might also be good to provide a `get_size()` function.

nordicjm commented:
Putting it here allows for users to "forget and not care" about it with a single reference that can be used for many instances. For example, with an API added on top of this for a boot mode API, I do not care if a user uses a prefix or checksum, that's up to them, I also have no clue how large their retention area is (e.g. SRAM could be 128 bytes, whilst the retention register on the nrf51 is a whopping 1 bytes) so all I want to do from that code is query the retention API: Is this data valid? If so, what boot mode should I be in? The retention driver is then responsible for ensuring the data is valid.

`get_size()` sounds better than `get_details()`

Laczen commented:
@nordicjm, why not leave the `is_valid` also to be defined by the user. A user that has only one byte would insert some parity check, a user that has 2kB might insert one CRC32 or multiple crc8 or even a hash/multiple hashes,... whatever suits the data that is shared.

IMO the added value is in the easy interface to a area that can be shared by multiple application (e.g. a bootloader and a application) and making the interface simple: `read/write/erase` and `getsize` as well as making the definition simple (just declaring the area) will stimulate users to take advantage of this interface. Adding predetermined functionality will require extra checks in the definition (e.g. avoid definition of a prefix and a crc in a one byte area) that can be avoided.

nordicjm commented:
I don't know if there is some stm32-centric device out there that requires 64-bit writes to RAM or something, I don't think they would have created such a device but then I never thought they'd have created an MCU that applies it to flash either.

Laczen commented:
> I don't know if there is some stm32-centric device out there that requires 64-bit writes to RAM or something, I don't think they would have created such a device but then I never thought they'd have created an MCU that applies it to flash either.

I wasn't thinking about something like that (if this was there it would be automatically set and wouldn't matter), I was thinking about a device that would have a separate register that would "manually" need to be written before it would even allow access to it. But then I also don't think this exists.

The biggest issue with the proposal I see is that it sets the validity check to a 1 byte value, this may not be acceptable for larger areas. Some check will need to be added to limit the size of the area in accordance with the used validity method, but this of course will limit the useability.

galak commented on dts/bindings/retention/retention-controller.yaml:
```diff
@@ -0,0 +1,20 @@
+# Common fields for data retention drivers
```
These properties need descriptions to understand what their purposes is.

nordicjm replied:
Added

galak commented on drivers/retention/Kconfig:
```diff
@@ -0,0 +1,28 @@
+menuconfig RETENTION
+	bool "Data retention support"
+	help
+	  Enables support for retaining data which can survive a reboot (and
```
I think this should be expanded to explain how these devices different from FLASH.

nordicjm replied:
Added

galak reviewed (dismissed):
base dts needs descriptions for properties, can't say if these properties make sense or not at this point.

carlescufi commented on include/zephyr/drivers/retention.h:
```diff
@@ -0,0 +1,198 @@
+ *
+ * @retval		0 on success else negative errno code.
+ */
+__syscall int32_t retention_read(const struct device *dev, size_t offset, uint8_t *buffer,
```
Use `retention_get` for consistency?

nordicjm replied:
Used read/write

carlescufi commented:
Architecture WG meeting:

- @bjarki-trackunit mentions that the validate and clear API calls do not map to hardware, so it might be better to leave them out of the driver and instead have them in a higher layer
- @nordicjm argues that it makes sense to have it in the driver
- @henrikbrixandersen states that the validate and clear are useful functions to have in common
- @galak agrees with @bjarki-trackunit and considers that the validation and clearing are out of the scope of this driver
- @galak states that we would have only a `size` property
- @bjarki-trackunit states that this API could replace the EEPROM API, and we expanded the scope of it to cover devices like EEPROM
- @nordicjm states that the mcp1 implementation is a battery-backed RAM one
- @henrikbrixandersen thinks that if this is going to be an API different from other, similar ones, we need to at least keep the `is_valid` call, because devices like battery-backed RAM. Those devices have a register that can be used to check if the data is valid.

bjarki-trackunit commented:
Just for reference:

### Battery backed ram
```
__subsystem struct bbram_driver_api {
	bbram_api_check_invalid check_invalid;
	bbram_api_check_standby_power check_standby_power;
	bbram_api_check_power check_power;
	bbram_api_get_size get_size;
	bbram_api_read read;
	bbram_api_write write;
};
```
### EEPROM
```
__subsystem struct eeprom_driver_api {
	eeprom_api_read read;
	eeprom_api_write write;
	eeprom_api_size size;
};
```

These can easily be either wrapped or replaced by the API in this PR. We loose the `check_power` adn `check_standby_power` from BBRAM, but note that only one of the 4 in-tree bbram drivers actually uses both those functions, the others use only check_invalid or none of them

nordicjm commented:
@yperess what is the use of `check_power` and `check_standby_power` for the npcx bbram? Are they needed/useful things to know for applications?

nordicjm commented:
Switched to using read/write for the API, reduced it down to the hardware level API as discussed in the architecture meeting and added support for the mcp device to use the power fail bit

bjarki-trackunit commented:
> I've seen one device do that and the whole idea of using an entire page of memory to store a boot state was nothing short of alarming.

EEPROM is writable and erasable byte-by-byte AFAK, so the difference really would only be that the "is valid" would either always return true or not be implemented for it

nordicjm commented:
> EEPROM is writable and erasable byte-by-byte AFAK, so the difference would only be that the "is valid" would either always return true or not be implemented for it, if i am correct that is

It is, however the other part of that application used the page number as an index, so the whole first page was wasted. There would be no "is valid" for it, eeproms do not support such a feature. @henrikbrixandersen what are your thoughts on moving eeprom drivers to this API? My thought is I don't want to do it, it might have the same functions but so does flash, you wouldn't put eeprom in flash and you wouldn't put retention in flash, they are separated because they are different classes of devices.

henrikbrixandersen commented:
> @henrikbrixandersen what are your thoughts on moving eeprom drivers to this API? My thought is I don't want to do it, it might have the same functions but so does flash, you wouldn't put eeprom in flash and you wouldn't put retention in flash, they are separated because they are different classes of devices.

I would be opposed to that. As you said, they are different device classes. One of the problems with combining things, that vaguely resemble each other, under same umbrella is that we loose the option to let these develop and mature independently of each other in the future. Also, the APIs would loose their semantics.

Laczen commented:
I would also oppose to trying to move the eeprom (or whatever device api) into the data retention API. IMHO this should be a kind op wrapper around whatever kind of menory, reserving a area on this memory and provide a simple way to write to and read from this reserved area. A `is_valid` function could be added only to provide support for devices that have this property (e.g. bbram). For the name maybe `rma` (reserved memory area) could be used.

nordicjm commented:
Some comments to discuss at today's architecture working group:

- Dominink has mentioned that there are devices which have a single "clear" command, e.g. https://ww1.microchip.com/downloads/en/DeviceDoc/MCP7951X-MCP7952X-Battery-Backed-SPI-RTCC-20002300C.pdf page 36 which has a "CLRRAM" command which clears all of the SRAM. So the function which was discounted back in the last architecture working group needs to be added back but only implemented for devices which support such a command (for this PR - none or only the zephyr RAM retention driver would support it)
- A capabilities function should probably be added
- Do the requirements for this make sense? If this is renamed to something generic e.g. nvmem - non-volatile memory, it could have a flash page added or something like the nordic UICR for devices which allow writes to all areas but each area only once then need to be cleared to allow writes again (or maybe not at all because they are programmed elsewhere) - if so, this would need to be communicated via the "capabilities" function
- I have moved the "software" part of this to here (no PR at present) https://github.com/nordicjm/zephyr/tree/bootmode - this subsystem probably needs to be discussed

Laczen commented:
To add to the discussion: Will this API replace somethings else ?
a. if not, there is no need to incorporate any special functionality from the "backend" device that is used. And thus no need 
for a `capabilities` and `is_valid` function, just read and write and the configuration is only a size and a "backend".
b. if yes, what will it replace ? Then we can look at the required functionalities.

My preference is with a.

de-nordic commented on include/zephyr/drivers/retention.h:
```diff
@@ -0,0 +1,172 @@
+ *
+ * @param dev		Retention device (area) to use.
+ *
+ * @retval		1 on success when data is valid, 0 when data is not valid, else negative
+ *			errno code.
```
What should callback return when the hardware does not implement the feature?

nordicjm replied:
This is defined in the design guidelines https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#return-codes

```
    APIs that are supported but not implemented shall return -ENOSYS.

    Optional APIs that are not supported by the hardware should be implemented and the return code in this case shall be -ENOTSUP.
```

mbolivar-nordic commented on drivers/retention/retention_nrf.c:
```diff
@@ -0,0 +1,105 @@
+	const struct nordic_nrf_retention_config *config = dev->config;
+	struct nordic_nrf_retention_data *data = dev->data;
+
+	if ((offset + size) > config->size) {
```
Probably worth using the overflow detection helpers from include/zephyr/sys/math_extras.h throughout.

nordicjm replied:
Those functions are for overflows of variable, this overflow check is for if e.g. someone tries reading 5 bytes from offset 6 in an area which is 10 bytes, so it overflows the area but not the variable type

nordicjm committed:
drivers: Add retained memory driver interface

Adds a new driver interface for retained memory devices which
can be used to store data and have it retained whilst the device is
powered thorugh different application execution states (though this
data may be lost in low power states).

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
drivers: retained_mem: Add nRF GPREGRET driver

Adds a driver for the Nordic nRF GPREGRET registers and adds
entries to the SoCs for this peripheral.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
drivers: retained_mem: Add RAM driver

Adds a non-initialised RAM-based retained memory driver.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented on drivers/nvmem/nvmem_zephyr_ram.c:
```diff
@@ -0,0 +1,135 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(zephyr_nvmem_ram, CONFIG_RETENTION_LOG_LEVEL);
```
missing rename

nordicjm commented on include/zephyr/drivers/nvmem.h:
```diff
@@ -0,0 +1,204 @@
+ *			what the data will be set to. Only implemented for hardware that
+ *			specifically supports this feature.
+ *
+ * @param dev		Retention device to use.
```
Missing rename

nordicjm commented on drivers/nvmem/Kconfig.zephyr:
```diff
@@ -0,0 +1,9 @@
+	default y
+	depends on DT_HAS_ZEPHYR_NVMEM_RAM_ENABLED
+	help
+	  Enable driver for storing retention data in RAM.
```
Missing rename

nordicjm committed:
tests: drivers: Add API test for retained_mem

Adds a test for the retained memory driver interface using GPREGRET
and RAM backends for the nrf52840dk_nrf52840 board and RAM backend
for qemu_cortex_m3 board.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.4: Add retained_mem driver

Adds details on the new retained memory driver.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

Laczen commented:
@nordicjm, the name nvmem non-volatile memory should only be used for something non-volatile. It seems like this name is chosen for something you would like it to behave as, not for it's actual behavior. A ram driver for nvmem should not be provided.

The name nvmem is also susceptible to misunderstanding and is way to broad for what this api is doing.

nordicjm committed:
MAINTAINERS: Add retained memory section and assign to self

Adds an entry for the retained memory area and assign myself to it.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
> the name nvmem non-volatile memory should only be used for something non-volatile. It seems like this name is chosen for something you would like it to behave as, not for it's actual behavior.

Indeed, most RAM would not be considered non-volatile, I have updated the Kconfig to say it is emulated and have text mentioning this

> A ram driver for nvmem should not be provided.

Disagree, it is a driver and it works, people are free to use it as they wish if they understand the implications of doing so.

> The name nvmem is also susceptible to misunderstanding and is way to broad for what this api is doing.

It provides a read and write API to non-volatile memory devices, there is no misunderstanding

nordicjm commented:
@erwango any issues or objections to moving the STM32 BBRAM driver to use this interface instead? Likewise @yperess for it8xxx2/npcx? Likewise @jvasanth1 for xec?

Laczen commented:
@nordicjm, taken from the introduction nvmem:
```
NVMEM is the abbreviation for Non Volatile Memory layer. It is used to retrieve configuration of SOC or 
Device specific data from non volatile memories like eeprom, efuses and so on.
```
So shouldn't it be used for eeprom ? At least if we are reusing the name we should inherit the meaning as well as its use.

nordicjm commented:
> So shouldn't it be used for eeprom ? At least if we are reusing the name we should inherit the meaning as well as its use.

No, this uses an *interface* which is *similar* to the linux NVMEM interface, it is not a copy of the linux driver, it is zephyr's own implementation and design of a non-volatile memory driver.

Laczen commented:
> > So shouldn't it be used for eeprom ? At least if we are reusing the name we should inherit the meaning as well as its use.
> 
> No, this uses an _interface_ which is _similar_ to the linux NVMEM interface, it is not a copy of the linux driver, it is zephyr's own implementation and design of a non-volatile memory driver.

It is confusing to reuse the name from a closely related system and give it a different meaning.

Laczen commented:
@nordicjm, TBH also in the linux kernel this could cause confusion. What about calling it `nvmemreg` where `reg` stands for region ? There would be no need to reuse this for a backend, but it could use whatever backend.

nordicjm commented:
I don't follow the importance of the "region" part - these drivers encompass the whole area of the underlying driver, they do not segment it into regions, the segmentation and region part is done by the subsystem which is an entirely different things/PR: https://github.com/zephyrproject-rtos/zephyr/pull/55125

erwango commented:
> @erwango any issues or objections to moving the STM32 BBRAM driver to use this interface instead? 

I was reading the whole discussion, wondering what the hell was the difference with bbram, until I found this question :-).
I'll have a look in details, but at first sight no objections.

Laczen commented:
@nordicjm, is the `is_valid` function required ? Can't this be incorporated in the `read` function.

nordicjm commented:
> is the is_valid function required ?

Only if the hardware supports it

> Can't this be incorporated in the read function.

No.

Laczen commented:
After going through this PR again it seems like this is introducing a (garbage  ?) driver for all nvmem things that do not have a driver yet. GPREGRET, bbram, and probably will be extended to efuse. Next to the generic ram backend probably backends will be added for eeprom, ...
It does not seem thoroughly tought trough as most of the things that are receiving a driver do not have a `is_valid` property. It feels like this is created as a driver for bbram while instead it should be using the bbram driver.

mbolivar-nordic commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,180 @@
+ */
+__syscall int32_t retained_mem_size(const struct device *dev);
+
+static inline int32_t z_impl_retained_mem_size(const struct device *dev)
```
IMO this should be ssize_t, not int32_t

nordicjm replied:
Fixed

mbolivar-nordic commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,180 @@
+	/* Validate user-supplied parameters */
+	if (size == 0) {
+		return -EINVAL;
+	} else if ((offset + size) > api->size(dev)) {
```
missing overflow check?

mbolivar-nordic replied:
Sorry if I am still misunderstanding the math_extras.h API, but shouldn't this be simplified to:

```c
	size_t end;
	bool overflow;

	/* Validate user-supplied parameters */
	if (size == 0) {
		return -EINVAL;
	}
	overflow = size_add_overflow(offset, size, &end);
	if (overflow || end > api->size(dev)) {
		return -EINVAL;
	}
```

?

Same comment applies elsewhere.

mbolivar-nordic commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,180 @@
+	/* Validate user-supplied parameters */
+	if (size == 0) {
+		return -EINVAL;
+	} else if ((offset + size) > api->size(dev)) {
```
overflow?

nordicjm replied:
Added

mbolivar-nordic commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,180 @@
+__syscall int32_t retained_mem_write(const struct device *dev, size_t offset,
+				     const uint8_t *buffer, size_t size);
+
+static inline int32_t z_impl_retained_mem_write(const struct device *dev, size_t offset,
```
Shouldn't this be "plain" int?

Nonblocking: are you sure that you will not need to support devices which report partial writes?

nordicjm replied:
This is a result code, it does not return the number of bytes written, picked int32_t as that is the result code size. For this API, do not want to support partial writes, can be see as a differentiator from e.g. flash API, devices should have byte level granularity for reads/writes.

mbolivar-nordic replied:
> This is a result code, it does not return the number of bytes written

I know -- if it was a result code or number of bytes I would have said `ssize_t`.

> picked int32_t as that is the result code size.

I'm saying that `int` would seem to be a better choice for 64-bit architectures to use the natural machine word size for the return value everywhere.

Same comment applies to the read and clear APIs.

> For this API, do not want to support partial writes, can be see as a differentiator from e.g. flash API, devices should have byte level granularity for reads/writes.

ACK.

nordicjm replied:
Fixed

mbolivar-nordic commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,180 @@
+__syscall int32_t retained_mem_read(const struct device *dev, size_t offset, uint8_t *buffer,
+				    size_t size);
+
+static inline int32_t z_impl_retained_mem_read(const struct device *dev, size_t offset,
```
"Plain" int, no?

Nonblocking: Did you consider the possibility of partial reads?

nordicjm replied:
As above

nordicjm replied:
Fixed

mbolivar-nordic commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,180 @@
+ */
+__syscall int32_t retained_mem_clear(const struct device *dev);
+
+static inline int32_t z_impl_retained_mem_clear(const struct device *dev)
```
"plain" int?

mbolivar-nordic commented on drivers/retained_mem/retained_mem_nrf_gpregret.c:
```diff
@@ -0,0 +1,147 @@
+#endif
+
+struct nrf_gpregret_config {
+	uint32_t addr;
```
Nonblocking: wouldn't it be simpler to declare this as a `uint8_t*`?

nordicjm replied:
Fixed

mbolivar-nordic commented on drivers/retained_mem/retained_mem_nrf_gpregret.c:
```diff
@@ -0,0 +1,147 @@
+			      CONFIG_RETAINED_MEM_INIT_PRIORITY,		\
+			      &nrf_gpregret_api);
+
+#ifdef CONFIG_MULTITHREADING
+DT_INST_FOREACH_STATUS_OKAY(NRF_GPREGRET_DEVICE_MULTITHREADING)
+#else
+DT_INST_FOREACH_STATUS_OKAY(NRF_GPREGRET_DEVICE)
+#endif
```
Nonblocking: you could simplify this a bit by:

- moving the `nrf_gpregret_data` definition into an `IF_ENABLED(CONFIG_MULTITHREADING, ...)`
- using COND_CODE_1 to pass either data or NULL to DEVICE_DT_INST_DEFINE

 See spi_nrfx_spim.c for examples.

nordicjm replied:
Fixed

mbolivar-nordic commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -42,6 +42,14 @@
+			gpregret1: gpregret1@4000051c {
+				compatible = "nordic,nrf-gpregret";
+				reg = <0x4000051c 0x1>;
+				status = "okay";
```
Do you really mean to enable the driver by default?

nordicjm replied:
Driver is marked as OK but will not be enabled in builds by default unless CONFIG_RETENTION is enabled, this is how it is with other peripherals e.g. i2c, spi, etc.

mbolivar-nordic commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -42,6 +42,14 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			gpregret1: gpregret1@4000051c {
```
Missing entries in validate_base_addresses.c for these devices

nordicjm replied:
Added

mbolivar-nordic commented on drivers/retained_mem/retained_mem_zephyr_ram.c:
```diff
@@ -0,0 +1,149 @@
+			      CONFIG_RETAINED_MEM_INIT_PRIORITY,				\
+			      &zephyr_retained_mem_ram_api);
+
+#ifdef CONFIG_MULTITHREADING
+DT_INST_FOREACH_STATUS_OKAY(ZEPHYR_RETAINED_MEM_RAM_DEVICE_MULTITHREADING)
+#else
+DT_INST_FOREACH_STATUS_OKAY(ZEPHYR_RETAINED_MEM_RAM_DEVICE)
+#endif
```
Similar nonblocking suggestion for how to simplify this applies here too

nordicjm replied:
Fixed

nordicjm committed:
doc: Add retained memory API and mark as experimental

Adds the the retained memory API to the list and marks it as
experimental.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,180 @@
+/**
+ * @brief Retained memory memory driver API
+ * API which can be used by a device to store data in a retained memory area. Retained memory is
+ * memory that is retained for the full power cycle of a device but is lost when power to the
```
Fixed

nordicjm commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,180 @@
+#endif
+
+/**
+ * @brief Retained memory driver interface
```
Added as experimental

nordicjm commented on drivers/retained_mem/retained_mem_handlers.c:
```diff
@@ -0,0 +1,38 @@
+						const uint8_t *buffer, size_t size)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_RETAINED_MEM));
+	return z_impl_retained_mem_write(dev, offset, buffer, size);
```
Fixed

nordicjm commented on drivers/retained_mem/retained_mem_handlers.c:
```diff
@@ -0,0 +1,38 @@
+					       uint8_t *buffer, size_t size)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_RETAINED_MEM));
+	return z_impl_retained_mem_read(dev, offset, buffer, size);
```
Fixed

nordicjm committed:
soc: arm: nordic: Add GPREGRET register validation

Adds validation for Nordic nRF GPREGRET registers.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

mbolivar-nordic commented on doc/hardware/peripherals/retained_mem.rst:
```diff
@@ -0,0 +1,24 @@
+
+The retained memory driver API provides a way of reading from/writing memory
+areas whereby the contents of the memory is retained whilst the device is
+powered (data may be list in low power modes).
```
s/list/lost/

nordicjm replied:
Fixed

mbolivar-nordic commented on drivers/retained_mem/retained_mem_nrf_gpregret.c:
```diff
@@ -0,0 +1,133 @@
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	memcpy(buffer, (void *)(config->addr + offset), size);
```
nonblocking: you can drop the cast now that you've changed `addr` to `uint8_t*`

nordicjm replied:
Can't reply to the above comment for some reason but `int` is not 64-bit on 64-bit builds, `long` is a 64-bit value:
```
printk("int = %d\n", sizeof(int));
```
=
```
Hello World! native_posix_64
int = 4
```

mbolivar-nordic commented on drivers/retained_mem/retained_mem_nrf_gpregret.c:
```diff
@@ -0,0 +1,133 @@
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	memcpy((void *)(config->addr + offset), buffer, size);
```
ditto

nordicjm replied:
Fixed

Laczen commented:
@nordicjm, a general remark: in other api's the `size==0` is not considered a error but is returning early with return value 0.

@mbolivar-nordic, regarding the overflow detection, I agree it could be done this way, but `size_t` is not a good type for something that is (or will become) a address. Maybe it would be better to define addresses as `unsigned long` and to provide  a general macro `bool in_range(unsigned long address, size_t size, size_t range)` this could be used for all devices that are using a `offset` (e.g. flash, eeprom, disk, ...). Something for the architecture meeting (@carlescufi)?

nordicjm commented:
> a general remark: in other api's the size==0 is not considered a error but is returning early with return value 0.

Updated to return 0 when size 0 is provided.

Laczen commented:
@nordicjm, regarding the use of `size_t` for the offset and `int32_t` instead of `int` as return value: other API's e.g. eeprom, flash are using `off_t` for offset and `int` as return value. Is there a reason not to follow this ?

nordicjm commented:
Can switch to `off_t` for offset. I don't like `int` because the size is not fixed, it's either a 16 or 32-bit variable, it might be fixed at 32-bit in the zephyr SDK

nordicjm commented:
Switched to off_t and int

Laczen commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	overflow = size_add_overflow(offset, size, &end);
+
```
This can no longer be used for a `off_t` type offset, this is signed

nordicjm replied:
Switched back to size_t, I do not fancy waiting another long period of time for something that might or might not be decided.

Laczen replied:
If it will not be decided, it will remain `off_t` and then so should this API. `size_t` is anyhow a bad choice, this is not a length.

nordicjm replied:
It is an off_t now

Laczen commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	overflow = size_add_overflow(offset, size, &end);
+
```
This can no longer be used (`off_t` is signed)

de-nordic commented on drivers/retained_mem/Kconfig.nrf:
```diff
@@ -0,0 +1,10 @@
+
+config RETAINED_MEM_NRF_GPREGRET
+	bool "nRF GPREGRET driver"
+	default y
```
Why do you want to have it y by default?

nordicjm replied:
This is how all drivers work, if you build for an nrf52840dk and enable i2c, it will automatically select the nordic i2c driver

Laczen commented on drivers/retained_mem/retained_mem_handlers.c:
```diff
@@ -0,0 +1,40 @@
+}
+#include <syscalls/retained_mem_size_mrsh.c>
+
+static inline int z_vrfy_retained_mem_read(const struct device *dev, off_t offset,
```
@de-nordic, the same is valid for `flash`, `eeprom` and probably others. I would propose to keep all at `off_t` and move the discussion wether this should be `unsigned long` or `....` to the architecture meeting. After this has been agreed upon all improper uses of `off_t` can be removed.

de-nordic replied:
OK, agree.

nordicjm replied:
Indeed, moved back to size_t

mbolivar-nordic commented:
On Mon, Feb 27 2023, Jamie wrote:
> Can't reply to the above comment for some reason but `int` is not
> 64-bit on 64-bit builds, `long` is a 64-bit value:

It seems... a little extreme to assert that an RTOS like zephyr will
never face an ILP64 target, given all the oddball architectures we
already support, no?

nordicjm commented on drivers/retained_mem/retained_mem_nrf_gpregret.c:
```diff
@@ -0,0 +1,132 @@
+
+LOG_MODULE_REGISTER(retained_mem_nrf_gpregret, CONFIG_RETAINED_MEM_LOG_LEVEL);
+
+#ifdef CONFIG_MULTITHREADING
+struct nrf_gpregret_data {
+	struct k_mutex lock;
+};
+#endif
```
It does - you cannot build this in mcuboot unless multithreading is enabled if you remove the #ifdef protection

nordicjm commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,194 @@
+ *
+ * @param dev		Retained memory device to use.
+ *
+ * @retval		0 on success else negative errno code.
```
Note on line 73:
```
 * Note that drivers must implement all functions, none of the functions are optional.
```

nordicjm commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,194 @@
+ * @brief	Callback API to get size of retained memory area.
+ * See retained_mem_size() for argument description.
+ */
+typedef ssize_t (*retained_mem_size_api)(const struct device *dev);
```
Just for errors, both drivers here cannot return an error because they're direct memory or registers but there might be a driver in the future that has a communication method that is not guaranteed to work e.g. IPC or I2C, so left the possibility of returning an error if it cannot be checked

nordicjm commented:
> It seems... a little extreme to assert that an RTOS like zephyr will never face an ILP64 target, given all the oddball architectures we already support, no?

I wasn't saying that it would never have one just that at present "int" is 32-bit regardless of architecture in zephyr. As a side note, and this was very quickly noticed when x86_64 CPUs first came out is that for most ALUs, simple 64-bit arithmetic is slower than 32-bit arithmetic, and also prevents SIMD for no real gain.

nordicjm commented on drivers/retained_mem/retained_mem_handlers.c:
```diff
@@ -0,0 +1,40 @@
+					   uint8_t *buffer, size_t size)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_RETAINED_MEM));
+	Z_OOPS(Z_SYSCALL_MEMORY_READ(buffer, size));
```
Good observation I missed, fixed

de-nordic commented on drivers/retained_mem/retained_mem_zephyr_ram.c:
```diff
@@ -0,0 +1,139 @@
+
+	zephyr_retained_mem_ram_lock_take(dev);
+
+	memcpy(buffer, (config->address + offset), size);
```
Before `memcpy` is called, the `config->address + offset` and `size` should be checked if they are within allowed range of the allowed "retained ram", otherwise you could use the call to write anywhere within address space.

de-nordic replied:
My bad, it is checked in syscall.

de-nordic commented on drivers/retained_mem/retained_mem_zephyr_ram.c:
```diff
@@ -0,0 +1,139 @@
+
+	zephyr_retained_mem_ram_lock_take(dev);
+
+	memcpy((config->address + offset), buffer, size);
```
With no `config->address` + offset and `size` check this can be used to exfiltrate any data accessible by kernel.

de-nordic commented on drivers/retained_mem/retained_mem_nrf_gpregret.c:
```diff
@@ -0,0 +1,136 @@
+
+	nrf_gpregret_lock_take(dev);
+
+	memcpy(buffer, (config->addr + offset), size);
```
`offset` (and size) needs to be checked to prevent gaining unrestricted access to RAM with the use of the API, from the user space.

de-nordic commented on drivers/retained_mem/retained_mem_nrf_gpregret.c:
```diff
@@ -0,0 +1,136 @@
+
+	nrf_gpregret_lock_take(dev);
+
+	memcpy((config->addr + offset), buffer, size);
```
`offset` (and size) needs to be checked to prevent using the API for privileged writes.

de-nordic commented on tests/drivers/retained_mem/api/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
Missing `# Copyright (c) 2023 Nordic Semiconductor ASA`

nordicjm replied:
Fixed

de-nordic reviewed (dismissed):
Looks OK. Thanks!

erwango commented on doc/hardware/peripherals/retained_mem.rst:
```diff
@@ -0,0 +1,24 @@
+Overview
+********
+
+The retained memory driver API provides a way of reading from/writing to memory
+areas whereby the contents of the memory is retained whilst the device is
+powered (data may be lost in low power modes).
```
> to memory areas whereby the contents of the memory is retained whilst the device is
powered (data may be lost in low power modes).

Sorry to come back on this, but description just matches plain usual RAM, which doesn't exactly fit the notion of retention.. Any way to make it clearer ?

nordicjm replied:
Not sure how to clarify that, for example on nordic parts, GPREGRET is 1 or 2 registers (an area), the RAM driver would be an area, not sure about any other devices that would add a driver for this.

erwango replied:
Didn't you mention about persistence across resets ?
Is that a required property for those areas ?

nordicjm replied:
Resets yes: `whereby the contents of the memory is retained whilst the device is powered` i.e. when the device remains powered but the application reboots, not if power to the device is lost

gromero reviewed (changes requested):
typo: Retained

gromero commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,194 @@
+typedef int (*retained_mem_clear_api)(const struct device *dev);
+
+/**
+ * @brief Retained memory memory driver API
```
is memory duplicated here?

nordicjm replied:
Fixed

nordicjm commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,194 @@
+
+/**
+ * @brief Retained memory driver interface
+ * @defgroup retained_mem_interface Retamined memory driver interface
```
Fixed

Laczen commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,198 @@
+		return -EINVAL;
+	}
+
+	overflow = size_add_overflow((uint32_t)offset, size, &end);
```
shouldn't this be a cast to `size_t` instead of `uint32_t` ?

Laczen replied:
And can't this whole construct be simplified to:
```
if ((api->size(dev) < len) || ((api->size(dev) - size) < (size_t)offset)) {
        return -EINVAL;
}

return api->read(dev, offset, buffer, size);
```
If `api->read()` handles the 0 length gracefully even the check for for zero length could be removed. IMO the check for zero length and exiting early is an unnecessary construct as the early exiting should not occur and even if it does there is no reason to exit early.

nordicjm replied:
The current method works

Laczen replied:
Sure it does, but why use 10 lines of code when it can be done in 3 ?

nordicjm replied:
Readability, the same reason you don't compress a 100 line function into a single line unless you're making a minified version of a library whereby you want absolutely no debugging ability.

Laczen replied:
Regarding readability this is certainly not an example of what I would call readability. In order to understand it it is required to grep in header files to find out what `size_add_overflow` is and what it's arguments are. Finally only to be left with the question "Why has the author made this so difficult ? It is only needed to check if the box is large enough and that the starting position is not to large."

Laczen commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,198 @@
+		return -EINVAL;
+	}
+
+	overflow = size_add_overflow((uint32_t)offset, size, &end);
```
cast to `size_t` ?

nordicjm replied:
Fixed

de-nordic commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,198 @@
+	bool overflow;
+
+	/* Validate user-supplied parameters */
+	if (size == 0) {
+		return 0;
+	} else if (offset < 0) {
+		return -EINVAL;
+	}
+
+	overflow = size_add_overflow((size_t)offset, size, &end);
+
+	if (overflow || end > api->size(dev)) {
+		return -EINVAL;
+	}
```
This part requires more time to figure out than probably needed.
Why not just follow the normal way (proposed by @Laczen, thanks):
```
 if ((dev->size < len) || ((dev->size(dev) - len) < (size_t)offset)) {
                return -EINVAL;
        }
```

I do not even think that usage of `size_add_overflow`, IMHO this thing should only by used on things that can be directly accessed by indexing address space.
We have to use cast to even use this thing on offset, which may not be save.
Consider a case when `sizeof(off_t) > sizeof(size_t)`: this is explicit cast, so no warning will be reported that `off_t` type variable gets cut down to `sizeof(size_t)`. This  means that in scenarios where `(size_t)(offset) < size`, call to `size_add_overflow` will report no overflow and `end` will be in range of `api->size(dev)`, which means that `api->read(...)` will be called with `offset` that is out of range.

nordicjm replied:
This is how they are set in the arm zephyr sdk:
```
typedef long _off_t;
typedef unsigned long __size_t;
```
So off_t can never be larger than size_t, only smaller, and the above check exits out if the provided offset is less than 0.

de-nordic replied:
The `off_t` is not defined within C standard, so far, and comes from POSIX definition. `size_t` is architecture dependent because it is supposed to be able to give size of largest memory object that can be addressed.
There is no dependency between `off_t` and `size_t` because they come from different worlds.
(and my proposed code above has risk of truncation error too, due to size_t cast).

I seen no benefit of using `size_and_overflow` when:
 1) `size_and_overflow` using code is just ugly now and looks jagged;
 2) it is actually half working because it will work as long as `off_t` <= `size_t` and the hides potential problem, as used `offset` is not actually the one that is used for overflow check.

nordicjm commented:
Updated, the discussion about size of `off_t` vs size of `size_t` doesn't really make any sense for this, `size_t` is either 32-bit or 64-bit, then from the GCC manual https://www.gnu.org/software/libc/manual/html_node/File-Position-Primitive.html:
```

Data Type: off_t

    This is a signed integer type used to represent file sizes. In the GNU C Library, this type is no narrower than int.

    If the source is compiled with _FILE_OFFSET_BITS == 64 this type is transparently replaced by off64_t. 

Data Type: off64_t

    This type is used similar to off_t. The difference is that even on 32 bit machines, where the off_t type would have 32 bits, off64_t has 64 bits and so is able to address files up to 2^63 bytes in length.

    When compiling with _FILE_OFFSET_BITS == 64 this type is available under the name off_t. 
```
Unless @stephanosio says there are plans to start adding this flag to the sdk-ng toolchains, it seems like a non-issue.

Laczen commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,186 @@
+	} else if (offset < 0) {
+		return -EINVAL;
+	} else if (((size_t)offset + size) > api->size(dev)) {
+		return -EINVAL;
```
@nordicjm, this can overflow, the proposal I made cannot.

nordicjm replied:
Fixed

de-nordic commented on include/zephyr/drivers/retained_mem.h:
Somewhere within this header please add single BUILD_ASSERT or other compile time check for `sizeof(off_t) <= size(size_t)`, to make sure that the `(size_t)` cast does not make any truncation that we miss while passing offset to API call.

de-nordic commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,196 @@
+
+	area_size = api->size(dev);
+
+	if (size > area_size || (area_size - size) < (size_t)offset) {
+		return -EINVAL;
+	}
```
Move this up and concatenate with `offset < 0` check.

nordicjm replied:
Done

de-nordic commented on include/zephyr/drivers/retained_mem.h:
```diff
@@ -0,0 +1,196 @@
+	size_t area_size;
+
+	/* Validate user-supplied parameters */
+	if (size == 0) {
+		return 0;
+	} else if (offset < 0) {
```
Make these separate checks, no need for `else if`.

de-nordic reviewed (approved):
Looks OK. Thanks!

@Laczen Thanks for reviewing.
